File: java-boot/src/test/java/se/callistaenterprise/kotlindsl/testutil/DslIntegrationTest.java
package se.callistaenterprise.kotlindsl.testutil;

import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.testcontainers.containers.PostgreSQLContainer;

public abstract class DslIntegrationTest {
    public static final String DBNAME = "orderdsl";
    public static final String DBUSERNAME = "test";
    public static final String DBPASSWORD = "test";
    public static final String CONTAINER_VERSION = "postgres:14";

    private static final PostgreSQLContainer container;

    static {
        container = new PostgreSQLContainer<>(CONTAINER_VERSION)
                .withDatabaseName(DBNAME)
                .withDatabaseName(DBUSERNAME)
                .withPassword(DBPASSWORD);
        container.start();
    }

    @DynamicPropertySource
    static void properties(DynamicPropertyRegistry registry) {
        registry.add("spring.datasource.url", container::getJdbcUrl);
        registry.add("spring.datasource.username", container::getUsername);
        registry.add("spring.datasource.password", container::getPassword);
    }
}

File: kotlin-boot/src/test/kotlin/se/callistaenterprise/kotlindsl/testutil/DslIntegrationTest.kt
package se.callistaenterprise.kotlindsl.testutil

import org.springframework.boot.test.util.TestPropertyValues
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.test.context.ContextConfiguration
import org.testcontainers.containers.PostgreSQLContainer

const val DBNAME = "orderdsl"
const val DBUSERNAME = "test"
const val DBPASSWORD = "test"

const val CONTAINER_VERSION = "postgres:14"

@ContextConfiguration(initializers = [DslIntegrationTest.Initializer::class])
abstract class DslIntegrationTest {

    companion object {
        val container = PostgreSQLContainer<Nothing>(CONTAINER_VERSION)
            .apply {
                withDatabaseName(DBNAME)
                withUsername(DBUSERNAME)
                withPassword(DBPASSWORD)
            }
    }

    internal class Initializer : ApplicationContextInitializer<ConfigurableApplicationContext> {

        override fun initialize(configurableApplicationContext: ConfigurableApplicationContext) {
            container.start()
            println("Starting testcontainer on port ${container.firstMappedPort}")
            TestPropertyValues.of(
                "spring.datasource.url=jdbc:postgresql://${container.host}:${container.firstMappedPort}/$DBNAME",
                "spring.datasource.username=$DBUSERNAME",
                "spring.datasource.password=$DBPASSWORD",
            ).applyTo(configurableApplicationContext.environment)
        }
    }

}
